### PR TITLE
Provide a builder for ListingOptions

### LINKED ISSUE #4178: Improve ergonomics of creating `ListingOptions`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Right now, `ListingOptions` is a public struct with some required fields and some that have reasonable default values

This means that to create one you need to provide all fields, which means that adding new fields requires changes to many places in DataFusion as well as downstream creates -- most recently https://github.com/apache/arrow-datafusion/issues/4177

For example, here is how you have to create one today:
```rust
    ListingOptions {
            format: Arc::new(file_format),
            collect_stat: false,
            file_extension: self.file_extension.to_owned(),
            target_partitions,
            table_partition_cols: self.table_partition_cols.clone(),
            file_sort_order: None,
        }
    }
```

**Describe the solution you'd like**
What I would like is a builder style API for this struct such that `new()` takes the required arguments and there are `with_` functions to set them

```rust

ListingOptions::new(file_format)
  .with_collect_statistics(false)
  .with_file_extension(self.file_extension.to_owned()),
  .with_target_partitions(target_partitions)
  .with_table_table_partition_cols(self.table_partition_cols.clone())
```

**Describe alternatives you've considered**

**Additional context**
Flagging this as good first issue as it is a software engineering effort that can be mostly guided by the compiler and would be a good way to get familiar with the codebase

cc @tustvold

### PR DESCRIPTION
**Pull request by mvanschellebeeck**
# Which issue does this PR close?

Closes #4178

# What changes are included in this PR?

Improves the ergonomics of [Listing Options](https://github.com/apache/arrow-datafusion/blob/a9add0e267f387559e155e9a477eee2cc99d0f2b/datafusion/core/src/datasource/listing/table.rs#L202) by providing a builder.

### CODE REVIEW

File: datafusion-examples/examples/parquet_sql_multiple_files.rs [hunk 1]
```diff
@@ -33,13 +33,11 @@ async fn main() -> Result<()> {
...
-        format: Arc::new(file_format),
-        table_partition_cols: vec![],
-        collect_stat: true,
-        target_partitions: 1,
-    };
+    let listing_options = ListingOptions::new(Arc::new(file_format))
+        .with_file_extension(FileType::PARQUET.get_ext())
+        .with_table_partition_cols(vec![])
+        .with_collect_stat(true)
+        .with_target_partitions(1);
```
Comment (alamb) [inline datafusion-examples/examples/parquet_sql_multiple_files.rs:40]: We could further  simplify this code and avoid explicitly specifying the default values (last three lines)

But that can also be done as a follow on PR

File: datafusion/core/src/datasource/listing/table.rs [hunk 2]
```diff
@@ -238,6 +238,73 @@ impl ListingOptions {
...
+    ///     .with_collect_stat(true);
+    ///
+    /// assert_eq!(listing_options.collect_stat, true);
+    /// ```
+    pub fn with_collect_stat(mut self, collect_stat: bool) -> Self {
+        self.collect_stat = collect_stat;
+        self
+    }
+
+    /// Set number of target partitions on [`ListingOptions`] and returns self.
```
Comment (alamb) [inline datafusion/core/src/datasource/listing/table.rs:293]: ❤️

**Review by alamb — APPROVED** [review]
Thank you @mvanschellebeeck  -- I think it looks great!

I'll leave this PR open for a day or two so others can comment if they want

cc @tustvold

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @mvanschellebeeck and @alamb

**Comment by alamb** [comment]
Thanks again @mvanschellebeeck  -- I took the liberty of merging up from master to resolve the conflicts (that I caused with https://github.com/apache/arrow-datafusion/commit/d2814c960168b45c4a0f5d7bbb72d9f412cb08bd / https://github.com/apache/arrow-datafusion/pull/4170)

**Comment by alamb** [comment]
I couldn't push directly to this PR 

```
$ git  push
Enumerating objects: 63, done.
Counting objects: 100% (63/63), done.
Delta compression using up to 16 threads
Compressing objects: 100% (27/27), done.
Writing objects: 100% (27/27), 3.26 KiB | 41.00 KiB/s, done.
Total 27 (delta 22), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (22/22), completed with 16 local objects.        
To github.com:mvanschellebeeck/arrow-datafusion.git
 ! [remote rejected]     mvanschellebeeck_master -> mvanschellebeeck_master (permission denied)
error: failed to push some refs to 'github.com:mvanschellebeeck/arrow-datafusion.git'

```

But I put my proposed changes in 
https://github.com/apache/arrow-datafusion/pull/4207

**Comment by alamb** [comment]
Thanks again @mvanschellebeeck  -- this is so much nicer

**Pull request closed without merging**